Fix GiEX QT06 water valve units

drakhart opened the pull request:
I made some fixes and improvements to the GiEX QT06 water valve:
- Replaced minutes with seconds. I realized that this device measures in seconds as soon as I set 2 irrigation cycles of 1 minute each with 1 minute interval and it just went crazy with instant on and offs. I set 60 "minutes" instead and it behaved as expected.
- Adjusted the max values to `3600` (seconds in an hour) because the existing `1440` (minutes in a day) didn't make sense anymore and setting `86400` (seconds in a day) just sounded ridiculous to me.
- Spelled `litres` instead of `Litres` as uppercase is just for the abbreviation (L).
- Removed some unnecessary units, especially the incorrect `GMT+8` for times as the device uses GMT (Zulu time).
- Improved some descriptions, especially those referring to last irrigation info.

Koenkk commented:
Thanks!

dgaust commented:
Is it 100% certain that the values for time are seconds? I just received this device, and setting the run time in seconds (i.e 2 seconds) results in the device running for 2 minutes.

The UI in the app only allows setting duration as minutes. And to confirm, when you set the duration in the app, the start and finish time is adjusted in minutes and not seconds. From the UI, I set the duration to 1s and the start and end time is exactly a minute apart.

![image](https://user-images.githubusercontent.com/3413862/211145150-d7ebfe9b-491a-499f-a532-0d733236d25b.png)

And then on completion of the sequence, the last irrigation duration is shows minutes

![image](https://user-images.githubusercontent.com/3413862/211145213-c59b399a-e0cf-4424-8277-5e50507c28e8.png)

drakhart commented:
I have four of these valves at home, and all four take the time value in seconds. They were bought in batches of two from [this store](https://es.aliexpress.com/item/1005004221200643.html). As a side note, they have never been paired with any app other than zigbee2mqtt.

Here are some screenshots as proof:
![image](https://user-images.githubusercontent.com/14996002/211173321-c266dec3-22aa-4cc1-bf2e-f95ab9f0c239.png)
![image](https://user-images.githubusercontent.com/14996002/211173338-05e51596-84e6-436a-91db-fa3f301d4b00.png)
![image](https://user-images.githubusercontent.com/14996002/211173353-8504630e-8ae9-471d-a4b5-2d44ebc0b9d6.png)
![image](https://user-images.githubusercontent.com/14996002/211173363-be927ae7-3c28-43d9-861a-de87a432d2db.png)

That being said, I don't doubt what you're saying so let's try and find a logic explanation for this difference:
- Did you buy yours from a different vendor?
- Could it be that some batches of units display a different behavior? Maybe based on serial number, I'll try to dig that info.
- Could it be that the firmware was updated after pairing them to the official app, and their behavior changed from then on?

dgaust commented:
I bought mine from Amazon https://www.amazon.com.au/dp/B0B1JN6KZX?ref=ppx_yo2ov_dt_b_product_details&th=1

I've only paired this device to zigbee2mqtt, and not any other app, so it has the firmware it was shipped with (it's not showing the firmware version in zigbee2mqtt either).

I've ordered a tuya gateway today, so once that arrives (probably wont be for a month), I'll hopefully be able to see more details and update any firmware if necessary.

tonimgp commented:
I need to modify the mcu and know in which gpio each component is connected.
Can you help me?

I want to use the device as a water meter at the entrance of the house

dgaust commented:
> I've only paired this device to zigbee2mqtt, and not any other app, so it has the firmware it was shipped with (it's not showing the firmware version in zigbee2mqtt either).
> 
> I've ordered a tuya gateway today, so once that arrives (probably wont be for a month), I'll hopefully be able to see more

Just circling back to this one, I received the gateway today and have paired it with the GeIX device. Unfortunately there are no firmware updates (currently on v1.06 for Zigbee and V1.0.0 for the MCU module), so I'm stuck with minutes as the most granual time I can set on the device.

drakhart commented:
I don't know what to say, I don't have any logic explanation as to why all my units work in seconds.... but since both the 1st commiter and you have your units in minutes maybe it's better to revert that specific change. Of course that's not gonna fix the time granularity for you, it's just that the UI will make sense.

dgaust commented:
I've ordered another one from aliexpress this time. Will see what happens when it arrives.

Knoedelauflauf commented:
I also received a new device today, ordered from amazon us. The device uses minutes instead of seconds, the granularity would be nice, but is not a hard requirement. May I help somehow?

stevewull commented:
I also received a new device today, ordered from Amazon US. It uses minutes.  Based on my reading of this thread and #13487, it appears that there are two versions:

1. Amazon US.  Uses Minutes. Manufacturer appears as:     _TZE200_sh1btabb

2. Aliexpress version. Uses seconds.  manufacturer appears as:  _TZE200_a7sghmms

(I'm guessing based on the posts...I do not have an aliexpress version to confirm.)  I plan to order some from ali in the future.

arroyorl commented:
Hello, how do you start the cycles. I've pair it with zigbee2mqtt, and I'm able to open/close the valve, but when I set values for irrigation target and interval, nothing happens. valve continues off (or on) indefinitelly. 

how to force to start the cycle?

thanks

regards

stevewull commented:
Hi arroyorl,

I have not found that either.  I use Node-Red (or a variety of different add-ins) within Home Assistant to kick off the start.






From: arroyorl ***@***.***>
Sent: Monday, February 13, 2023 6:13 AM
To: Koenkk/zigbee-herdsman-converters ***@***.***>
Cc: Stephen Wullschleger ***@***.***>; Comment ***@***.***>
Subject: Re: [Koenkk/zigbee-herdsman-converters] Fix GiEX QT06 water valve units (PR #4929)


Hello, how do you start the cycles. I've pair it with zigbee2mqtt, and I'm able to open/close the valve, but when I set values for irrigation target and interval, nothing happens. valve continues off (or on) indefinitelly.

how to force to start the cycle?

thanks

regards

—
Reply to this email directly, view it on GitHub<https://github.com/Koenkk/zigbee-herdsman-converters/pull/4929#issuecomment-1428006591>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALA37OKMNSNSUTCDDZ74IHDWXI6OJANCNFSM6AAAAAAR52Q46Y>.
You are receiving this because you commented.Message ID: ***@***.******@***.***>>

fedegrc commented:
Hi, my GiEX QT06 manufacturer `_TZE200_a7sghmms` works as described by @drakhart, the values for time are seconds. I bought it in https://www.aliexpress.com/item/1005004222098040.html
Following a picture of the hardware version:
![1677625180691](https://user-images.githubusercontent.com/14953828/222007060-fb713812-4736-468b-8b4e-024294acf25c.jpg)
How can I see the GiEX firmware version in Z2M?

WhimsySpoon commented:
I have a `_TZE200_sh1btabb` from AliExpress and it's being reported in seconds, but the `irrigation_start_time` and `irrigation_end_time` values are out. They appeared to have started at `00:00:00` at point of pairing and incremented since.

WhimsySpoon commented:
I have created an external converter using the previous version of `giex.js` and everything is now working as it should.

Koenkk commented:
@WhimsySpoon can you provide me the external converter? Then I can integrate the fix.

WhimsySpoon commented:
@Koenkk I literally just copied the contents of `giex.js` out of the "before" in the diff of this PR and created an external converter from it (attached as a text file).

I was wondering if we could create two files, the only difference being the device fingerprint. I know it would be a lot of duplicated code, but it would get the job done. If it's just a case of this, I'd be happy to raise the PR myself.

Also, is there anything we could do with the `irrigation_start_time` and `irrigation_end_time` timestamps to make them local?

[giex.txt](https://github.com/Koenkk/zigbee-herdsman-converters/files/10902351/giex.txt)

fedegrc commented:
> Removed some unnecessary units, especially the incorrect `GMT+8` for times as the device uses GMT (Zulu time).

@drakhart, remove the GMT+8 is ok for `_TZE200_a7sghmms`. However, my Z2M is showing my localtime, which is not GMT. Should we remove `(GMT)` the in the description?

`Last irrigation start time (GMT)` --> `Last irrigation start time`
`Last irrigation end time (GMT)` --> `Last irrigation end time`

Also, it would be great to have the date and time instead of just the time.

WhimsySpoon commented:
Is your local time GMT+8 by any chance? This is what seems to be hard-coded in the devices.

fedegrc commented:
> @fedegrc Is your local time GMT+8 by any chance? This is what seems to be hard-coded in the devices.

My local time is GMT-3.

fedegrc commented:
We have detected the following differences so far:

- `_TZE200_sh1btabb`: Uses Minutes. Timezone is GMT+8
- `_TZE200_a7sghmms`: Uses seconds. Timezone is local

Koenkk commented:
@drakhart do I understand correctly that different units are used across devies?

drakhart commented:
> @drakhart, remove the GMT+8 is ok for `_TZE200_a7sghmms`. However, my Z2M is showing my localtime, which is not GMT. Should we remove `(GMT)` the in the description?

IMO yes, it'd be better to have no suffix than having the wrong one.

> @drakhart do I understand correctly that different units are used across devices?

Apparently this is the case, according to the findings in this thread. Unfortunately for me all four units I own are `_TZE200_a7sghmms`, so I can't compare myself. Anyway it looks like I should have been more cautious before assuming that the original units were wrong and changing them.

stevewull commented:
I have three of the _TZE200_sh1btabb version.  Irrigation target is in minutes.

I live at GMT -08:00 Time zone (Americas/Pacific)  Although daylight savings time adjustment is a week away.

The irrigation_start_time seems to be GMT +08:00

Koenkk commented:
@drakhart I see, could you make a PR to fix the issues?

WhimsySpoon commented:
Could we also have something along the lines of, but not as hacky as the following to handle the timezone conversion?

```
//times from device hard-coded to GMT+8
let startTime = new Date(Date.parse("1900-01-01 " + value + " GMT+0800"));

return {irrigation_start_time: startTime.toTimeString().split(' ')[0]};
```

drakhart commented:
> @drakhart I see, could you make a PR to fix the issues?

Sure, I'll try my best.

> Could we also have something along the lines of, but not as hacky as the following to handle the timezone conversion?

I'll look into something, though I'm not sure I'll be able to find any solution less "hackier" than yours.

BTW, @fedegrc, are you absolutely sure about `_TZE200_a7sghmms` reporting your local GMT-3 time out of the box? I own four of those and all of them are reporting GMT/Zulu, in fact I'm doing the conversion to my local time (GMT+1) inside Home Assistant.

WhimsySpoon commented:
I have something working on my custom converter. I'll upload it tomorrow.

fedegrc commented:
> BTW, @fedegrc, are you absolutely sure about `_TZE200_a7sghmms` reporting your local GMT-3 time out of the box? I own four of those and all of them are reporting GMT/Zulu, in fact I'm doing the conversion to my local time (GMT+1) inside Home Assistant.

@drakhart, I took this screenshot about 12 minutes ago. My time zone is GMT-3. I'm using Z2M as an add-on in Home Assistant.

![1678314568916](https://user-images.githubusercontent.com/14953828/223867035-0eefc1cf-553d-46c0-b5f2-02fdff4054ee.jpg)

drakhart commented:
You're absolutely right, @fedegrc, my Ubuntu instance was misconfigured to GMT instead of GMT+1. Irrigation start and end times started showing the correct value as soon as I set the proper timezone system wide. It looks like we can completely omit the timezone suffix for `_TZE200_a7sghmms` and just assume that the value is in local time.

Can anyone with a `_TZE200_sh1btabb` confirm if these times are shown in their local timezone or GMT+8, please?

WhimsySpoon commented:
@drakhart I'm in the GMT time zone using a `_TZE200_sh1btabb` and mine are coming through as GMT+8. I have the following code in a custom converter to handle the conversion to GMT:

````
  case dataPoints.giexWaterValveIrrigationStartTime: {
      
      //times from device hard-coded to GMT+8
      let startTime = new Date(Date.parse("1900-01-01 " + value + " GMT+0800"));

      return {irrigation_start_time: startTime.toTimeString().split(' ')[0]};
  }
  case dataPoints.giexWaterValveIrrigationEndTime: {

      if (value=='--:--:--') return {irrigation_end_time: value};

      //times from device hard-coded to GMT+8
      let endTime = new Date(Date.parse("1900-01-01 " + value + " GMT+0800"));
      
      return {irrigation_end_time: endTime.toTimeString().split(' ')[0]};
  }
````

drakhart commented:
> We have detected the following differences so far:
> 
> * `_TZE200_sh1btabb`: Uses Minutes. Timezone is GMT+8
> * `_TZE200_a7sghmms`: Uses seconds. Timezone is local

Ok, so whenever I can I'll push a new PR based on this info by @fedegrc and implementing something on the lines of @WhimsySpoon's code for `_TZE200_sh1btabb` time conversion. Just bear with me a couple of days, because I'm going through a flu right now and my brain is completely fried.

WhimsySpoon commented:
Thanks @drakhart. No pressure to get this done, focus on your health and getting well! If there's anything I can do to assist with the code, please shout.

fedegrc commented:
In the next PR I would like the maximum irrigation time to be changed.

I think we should go back  from 3600 to 1440 to the max irrigation_target in the `_TZE200_sh1btabb` (uses minutes).

Regarding the `_TZE200_a7sghmms` (uses seconds), I did some tests to get the maximum `irrigation_target` value accepted and I found that it is 65535. We can put that value, or round to 43200 (12hs) or 64800 (18hs).

I can't confirm, but surely `cycle_irrigation_interval` also has a maximum of 65535.

fedegrc commented:
@koenkk, to integrate the water consumption sensor in the home-assistant energy dashboard, we need to add the following ([ha doc](https://www.home-assistant.io/docs/energy/water/)):

- `state_class: total_increasing`
- `device_class: water`

Can it be set in the device's herdsman-converters?

Koenkk commented:
@fedegrc this applies to the `water_consumed` sensor right?

fedegrc commented:
@Koenkk , yes, this applies to the `water_consumed` sensor.

drakhart commented:
New PR opened: https://github.com/Koenkk/zigbee-herdsman-converters/pull/5567

Please have a look, @fedegrc and @WhimsySpoon, to see if this addresses your suggestions/requests.

drakhart commented:
BTW, can anyone with the `_TZE200_sh1btabb` confirm if the box states "Thread: NH"? For `_TZE200_a7sghmms` it says "Thread: BSP". That makes me think that the `B` in the `QT06-B-Z-V3.0` motherboard pic shared by @fedegrc could stand for `QT06 BSP ZigBee 3.0`.

If it's like this we can give these two models more friendly names, like `QT06-BSP`/`QT06-NH` or just `QT06-B`/`QT06-N`.

WhimsySpoon commented:
@drakhart isn't the Thread to do with the pipe fitting? Unfortunately I don't still have the box.

https://arthurharris.com/news/npt-vs-bsp/

drakhart commented:
> @drakhart isn't the Thread to do with the pipe fitting? Unfortunately I don't still have the box.

@WhimsySpoon in theory yes, it should only have to do with the pipe fitting, but that `B` printed on `_TZE200_a7sghmms`'s motherboard is very sus. Would you mind opening the plastic cover of your `_TZE200_sh1btabb` unit and taking a look to make sure? Of course only if you are comfortable with it; I understand that not everyone is.

WhimsySpoon commented:
@drakhart  no problem at all. Is this what you're after?

![image](https://user-images.githubusercontent.com/19993138/224539943-abdd78c2-aac2-4644-8231-ca9a856decd1.png)

drakhart commented:
> @drakhart no problem at all. Is this what you're after?

Yes, thank you. Unfortunately that's exactly the same naming as on @fedegrc's `_TZE200_a7sghmms`, so I still don't know what does it make a difference for this device to have two different models.

WhimsySpoon commented:
I wonder if it refers to a different component, the valve for instance. But it does seem odd that there are software differences. I wonder if it's worth emailing GIEX to see if they'll explain.

drakhart commented:
I don't feel the need to: I think we're covered for now, thanks to you and the rest of people who reported info about the different variants.

WhimsySpoon commented:
I did actually message them shortly after my response above. The following is what I received a few hours ago. Based on their response around the min irrigation time, it looks like `_TZE200_a7sghmms` is the latest version.

I've replied and asked if they would be able to provide a firmware update to bring all devices in line with the latest

````
Hello. This is a different batch of product. The firmware is constantly being improved. One has a minimum irrigation time in minutes and one has a minimum irrigation time in 10 seconds.

The device in the warehouse now has a minimum irrigation time of 10s.
````

WhimsySpoon commented:
Unfortunately, they've just confirmed that the firmware **isn't** ugpradable on these devices.

drakhart commented:
> Maybe those 10 seconds are to avoid some mechanical problem in the valve. It might be better to change the minimum time to 10 seconds for `_TZE200_a7sghmms`. That is, that the possible values are 10, 11, 12, ..., 43200.

Yes, makes sense, if they updated the device to work like this they must have a good/valid reason. Unfortunately it's not as easy as setting the minimum value to `10`: we'd be breaking the `0` value which means "do not turn off until manually told to do so". Also we need to check if the device is in `duration` mode, as it shouldn't affect the `capacity` one. Tricky, but I'll try.

> If the `_TZE200_a7sghmms` is the latest version, it may be more consistent to swap the names:
> 
> `_TZE200_sh1btabb` --> `QT06_1` `_TZE200_a7sghmms` --> `QT06_2`

👍

drakhart commented:
Good news, @fedegrc: I managed to implement the minimum 10 seconds duration for `_TZE200_a7sghmms`.

WhimsySpoon commented:
As expected, they have confirmed that `_TZE200_a7sghmms` is the latest version of the two.

WhimsySpoon commented:
@Giexperience Is the device available to purchase on Ali Express yet? Could you share a link please?

Giexperience commented:
yeah. This is the sales link on aliexpress.
https://es.aliexpress.com/item/1005006547328147.html?gatewayAdapt=glo2esp#nav-moretolove


| |
GIEX-CINLIN-RAUL
|
|
***@***.***
|
---- Replied Message ----
| From | ***@***.***> |
| Date | 3/1/2024 20:35 |
| To | ***@***.***> |
| Cc | ***@***.***> ,
***@***.***> |
| Subject | Re: [Koenkk/zigbee-herdsman-converters] Fix GiEX QT06 water valve units (PR #4929) |

@Giexperience Is the device available to purchase on Ali Express yet? Could you share a link please?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

dgaust commented:
I ordered one to replace one of the earlier versions. Hopefully a simple tweak to integrate it

dgaust commented:
I received mine the other day, seems a lot more robust that the previous version and a fair bit smaller as well. I was able to get it running in zigbee2mqtt using the below in an external converter file

```
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const legacy = require('zigbee-herdsman-converters//lib/legacy');
const MINUTES_IN_A_DAY = 1440;
const SECONDS_IN_12_HOURS = 43200;

module.exports = [
    {
        fingerprint: [
            {modelID: 'TS0601', manufacturerName: '_TZE204_7ytb3h8u'},
        ],
        model: 'GX02',
        vendor: 'GiEX',
        description: 'GX02 GiEX Water Valve',
        extend: [],
        meta: {},
        onEvent: tuya.onEventSetLocalTime,
        fromZigbee: [legacy.fromZigbee.giexWaterValve],
        toZigbee: [legacy.toZigbee.giexWaterValve],
        exposes: [
            e.battery(),
            e.binary(legacy.giexWaterValve.state, ea.STATE_SET, 'ON', 'OFF')
                .withDescription('State'),
            e.enum(legacy.giexWaterValve.mode, ea.STATE_SET, ['Duration', 'Capacity'])
                .withDescription('Irrigation mode'),
            e.numeric(legacy.giexWaterValve.cycleIrrigationNumTimes, ea.STATE_SET)
                .withValueMin(0)
                .withValueMax(100)
                .withDescription('Number of cycle irrigation times, set to 0 for single cycle'),
            e.numeric(legacy.giexWaterValve.irrigationStartTime, ea.STATE)
                .withDescription('Last irrigation start time'),
            e.numeric(legacy.giexWaterValve.irrigationEndTime, ea.STATE)
                .withDescription('Last irrigation end time'),
            e.numeric(legacy.giexWaterValve.lastIrrigationDuration, ea.STATE)
                .withDescription('Last irrigation duration'),
            e.numeric(legacy.giexWaterValve.waterConsumed, ea.STATE)
                .withUnit('L')
                .withDescription('Last irrigation water consumption'),
            e.numeric(legacy.giexWaterValve.irrigationTarget, ea.STATE_SET)
                .withValueMin(0)
                .withValueMax(SECONDS_IN_12_HOURS)
                .withUnit('seconds or litres')
                .withDescription('Irrigation target, duration in seconds or capacity in litres (depending on mode), ' +
                    'set to 0 to leave the valve on indefinitely, ' +
                    'for safety reasons the target will be forced to a minimum of 10 seconds in duration mode'),
            e.numeric(legacy.giexWaterValve.cycleIrrigationInterval, ea.STATE_SET)
                .withValueMin(0)
                .withValueMax(SECONDS_IN_12_HOURS)
                .withUnit('sec')
                .withDescription('Cycle irrigation interval'),
        ],
    },
];
```

dgaust commented:
> yeah. This is the sales link on aliexpress. https://es.aliexpress.com/item/1005006547328147.html?gatewayAdapt=glo2esp#nav-moretolove | | GIEX-CINLIN-RAUL | | ***@***.*** | ---- Replied Message ---- | From | ***@***.***> | | Date | 3/1/2024 20:35 | | To | ***@***.***> | | Cc | ***@***.***> , ***@***.***> | | Subject | Re: [Koenkk/zigbee-herdsman-converters] Fix GiEX QT06 water valve units (PR #4929) | @Giexperience Is the device available to purchase on Ali Express yet? Could you share a link please? — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were mentioned.Message ID: ***@***.***>

The GX02 will be supported in the next release of Zigbee2MQTT

https://github.com/Koenkk/zigbee-herdsman-converters/issues/7229

WhimsySpoon commented:
Thanks for converter above, and submitting the code.

Have you noticed that when pressing the button on the GX02, the Irrigation End Time changes to `--:--:--` until it's closed again, but when changing the state via Zigbee, it automatically changes  to now, or now + irrigation target duration?

dgaust commented:
It's expected behaviour from what I can tell. When you turn it in using the physical button it's manual watering, and so doesn't have a duration or capacity set. It's on until you turn it off.

WhimsySpoon commented:
Ah, that makes sense. Thanks for the clarification.
